### PR TITLE
[engine] Refuse to downgrade optional field values which are not None 

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances #17082

- Principled recode of `importValue`/`SRecord` to handle UPGRADE/DOWNGRADE
- Refuse to downgrade optional field values which are not None
- Fix buglets in `ExplicitDisclosureTest` which triggered illegal downgrade case (thanks Carl)

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -1099,31 +1098,76 @@ private[lf] object Speedy {
...
-                    go(AstUtil.substitute(fieldType, subst), fieldValue)
+
+                // This code implements the compatibility transformation used for up/down-grading
+                // And handles the cases:
+                // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                val numS: Int = sourceElements.length
+                val numT: Int = targetFieldsAndTypes.length
+
+                // traverse the sourceElements, "get"ing the corresponding target type
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1115]: Nit pick: maybe `"getting"` might be better here?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 2]
```diff
@@ -1099,31 +1098,76 @@ private[lf] object Speedy {
...
+                          case Some(sourceField) =>
+                            // value is not normalized; check field names match
+                            assert(sourceField == targetField)
+                        }
+                        val typ: Type = AstUtil.substitute(targetFieldType, subst)
+                        val sv: SValue = go(typ, v)
+                        List(sv)
+                      }
+                      case None => { // DOWNGRADE
+                        assert(i >= numT)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1133]: Should we also assert that `numS > numT` - i.e. we are actually within a downgrade context?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -1099,31 +1098,76 @@ private[lf] object Speedy {
...
-                // TODO: https://github.com/digital-asset/daml/issues/17082
-                // - disallow dropping extra fields which have value other than None.
-                val actualValues = (lookupResult.dataRecord.fields.iterator zip fields.iterator)
-                  .map { case ((_, fieldType), (_, fieldValue)) =>
-                    go(AstUtil.substitute(fieldType, subst), fieldValue)
+
+                // This code implements the compatibility transformation used for up/down-grading
+                // And handles the cases:
+                // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1111]: Maybe also comment about the case where `numS == numT`? Simply worried that a quick read here might give the impression there are only 2 cases that are ever possible.

**Review by carlpulley-da — CHANGES_REQUESTED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1115]: It was deliberate. Refers to the `get` 4 lines down.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1133]: I don't think this assertion would add anything, 
`i`  is bound by `sourceElements.toSeq.zipWithIndex` and so ranges from 0 to `numS-1`  
So `i >= numT` implies `numS > numT`
I will add a comment to this effect.

Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1133]: Got you.

Yes, so when `i >= numT`, we know at leats one of our source indexes outstrips `numT`, and so can infer that `numS > numT`.

Yeah, a comment here would help a great deal. Thanks for that.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 3 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1111]: Extended commnet

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1139]: Actually. What the hell. I will add the assertion you suggested.

**Review by carlpulley-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 4]
```diff
@@ -1099,31 +1098,82 @@ private[lf] object Speedy {
...
+                // when there is no corresponding type, we must be downgrading, and so we insist the value is None
+                val values0: List[SValue] =
+                  sourceElements.toSeq.zipWithIndex.flatMap { case ((optName, v), i) =>
+                    targetFieldsAndTypes.get(i) match {
+                      case Some(x) => {
+                        val (targetField, targetFieldType): (Name, Type) = x
+                        optName match {
+                          case None => ()
+                          case Some(sourceField) =>
+                            // value is not normalized; check field names match
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1130]: This comment is a bit off. All value that pass through `imporValue` are actually normalized. 
Normalization of old version actually have field name.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 5]
```diff
@@ -1099,31 +1098,82 @@ private[lf] object Speedy {
...
+                // although it is still possible that the source and target types are different,
+                // but since we don't consult the source type (may be unavailable), we wont know.
+
+                val numS: Int = sourceElements.length
+                val numT: Int = targetFieldsAndTypes.length
+
+                // traverse the sourceElements, "get"ing the corresponding target type
+                // when there is no corresponding type, we must be downgrading, and so we insist the value is None
+                val values0: List[SValue] =
+                  sourceElements.toSeq.zipWithIndex.flatMap { case ((optName, v), i) =>
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1123]:
```suggestion
                  sourceElements.toSeq.view.zipWithIndex.flatMap { case ((optName, v), i) =>
```

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM. Thanks

**Pull request merged**